pysteps/blending/steps.py

```diff
@@ -65,1723 +64,2858 @@
+        self.__precip = self.__precip[-1, :, :]
+        print("Starting blended nowcast computation.")
 
-      darts/mch:
-      p_par  = [24.27562298, 0.11297186, -27.30087471]
-      p_perp = [-7.80797846e+01, -3.38641048e-02, 7.56715304e+01]
+        if self.__config.measure_time:
+            starttime_mainloop = time.time()
+        self.__state.extrapolation_kwargs = deepcopy(self.__config.extrapolation_kwargs)
+        self.__state.extrapolation_kwargs["return_displacement"] = True
 
-      darts/fmi+mch:
-      p_par  = [16.55447057, 0.14160448, -19.24613059]
-      p_perp = [14.75343395, 0.11785398, -16.26151612]
+        self.__state.precip_cascades_prev_subtimestep = deepcopy(
+            self.__state.precip_cascades
+        )
+        self.__state.cascade_noise_prev_subtimestep = deepcopy(
+            self.__state.precip_noise_cascades
+        )
 
-      lucaskanade/fmi:
-      p_par  = [2.20837526, 0.33887032, -2.48995355]
-      p_perp = [2.21722634, 0.32359621, -2.57402761]
+        self.__state.time_prev_timestep = [
+            0.0 for j in range(self.__config.n_ens_members)
+        ]
+        self.__state.leadtime_since_start_forecast = [
+            0.0 for j in range(self.__config.n_ens_members)
+        ]
 
-      lucaskanade/mch:
-      p_par  = [2.56338484, 0.3330941, -2.99714349]
-      p_perp = [1.31204508, 0.3578426, -1.02499891]
+        # iterate each time step
+        for t, subtimestep_idx in enumerate(self.__timesteps):
+            self.__determine_subtimesteps_and_nowcast_time_step(t, subtimestep_idx)
+            if self.__config.measure_time:
+                starttime = time.time()
+            self.__decompose_nwp_if_needed_and_fill_nans_in_nwp(t)
+            self.__find_nowcast_NWP_combination(t)
+            self.__determine_skill_for_current_timestep(t)
+            # the nowcast iteration for each ensemble member
+            final_blended_forecast_all_members_one_timestep = [
+                None for _ in range(self.__config.n_ens_members)
+            ]
 
-      lucaskanade/fmi+mch:
-      p_par  = [2.31970635, 0.33734287, -2.64972861]
-      p_perp = [1.90769947, 0.33446594, -2.06603662]
+            def worker(j):
+                # The state needs to be copied as a dataclass is not threadsafe in python
+                worker_state = deepcopy(self.__state)
+                self.__determine_skill_for_next_timestep(t, j, worker_state)
+                self.__determine_weights_per_component(worker_state)
+                self.__regress_extrapolation_and_noise_cascades(j, worker_state)
+                self.__perturb_blend_and_advect_extrapolation_and_noise_to_current_timestep(
+                    t, j, worker_state
+                )
+                # 8.5 Blend the cascades
+                final_blended_forecast_single_member = []
+                for t_sub in self.__state.subtimesteps:
+                    # TODO: does it make sense to use sub time steps - check if it works?
```

Comment:
Regarding the sub timesteps: I would love it if we could make it so that pysteps can just handle a variable timestep.

When you would go for example from a 5 to 15 minute timestep you could aggregate the previous 3 precipitation fields and multiply the last motion field by 3 and you would then probably just be able to continue in a 15 minute timestep from there right?

This would be much preferable to the sub timestep, which still has to do all the computations at the smaller timesteps. Also it should make the code easier to follow maybe

Reply:
I don't think this is part of the refactoring, but maybe something for after

Reply:
Indeed, but I think the refactoring makes the process a lot easier to to this.